lighthouse-core/lib/page-functions.js

```diff
@@ -98,19 +102,21 @@ function checkTimeSinceLastLongTask() {
  * @template {string} T
  * @param {T} selector Optional simple CSS selector to filter nodes on.
  *     Combinators are not supported.
- * @return {Array<HTMLElementByTagName[T]>}
+ * @return {Array<ParseSelector<T>>}
  */
 function getElementsInDocument(selector) {
   const realMatchesFn = window.__ElementMatches || window.Element.prototype.matches;
-  /** @type {Array<HTMLElement>} */
+  /** @type {Array<ParseSelector<T>>} */
   const results = [];
 
-  /** @param {NodeListOf<HTMLElement>} nodes */
+  /** @param {NodeListOf<Element>} nodes */
   const _findAllElements = nodes => {
     for (let i = 0, el; el = nodes[i]; ++i) {
       if (!selector || realMatchesFn.call(el, selector)) {
+        // @ts-expect-error - el is verified as matching above, tsc just can't verify it through the .call().
         results.push(el);
       }
```

Comment:
I thought it would be pretty sweet to also type `Element.prototype.matches` like

```ts
interface Element {
  matches<S extends string>(selectors: S): this is ParseSelector<S>;
}
```
and then `realMatchesFn.call(el, selector))` would verify that `el` could pushed into the `results` array, but unfortunately augmenting `matches` like that makes it so the method is overridden, and the old non-specific form ([`matches(selectors: string): boolean`](https://github.com/microsoft/TypeScript/blob/203a5ce78f1d9a378e8312e32956973f8334af8b/lib/lib.dom.d.ts#L5265) is the form that wins when `realMatchesFn.call` picks one of the overrides to use as a type (while calling it directly would pick the above, more specific version).

As a result, there's no type guard and so `el` remains an `Element`, which we already knew :)